content/various/FastInput.h

```diff
@@ -0,0 +1,32 @@
+/**
+ * Author: chilli
+ * License: CC0
+ * Source:
+ * Description: Returns an integer. Usage requires your program to pipe in
+ * input from file. Can replace calls to gc() with getchar_unlocked() if extra
+ * speed isn't necessary (60% slowdown).
+ * Status: tested on SPOJ INTEST
+ * Time: About 5x as fast as cin/scanf.
+ * Usage: ./a.out < input.txt
+ */
+#pragma once
+
+struct GC {
+	char buf[1 << 16 | 1];
+	int bc = 0, be = 0;
+	char operator()() {
+		if (bc >= be) {
+			be = fread(buf, 1, sizeof(buf) - 1, stdin);
+			buf[be] = bc = 0;
```

Comment:
The whole scheme with adding +1 to the buf size and zero-terminating this buffer is just to deal with case where fread returns 0, right? how about
```
buf[0] = bc = 0;
be = fread(buf, 1, sizeof(buf), stdin);
```
and making buf have size 1 << 16?

Reply:
The point is so that `gc()` returns 0 on EOF. I don't think that happens with this right?

Reply:
So in my version, if fread returns zero, we return the value of `buf[0]`, which is set to zero beforehand. This assumes that fread doesn't write past the size it returns, which I *think* should be true per spec, and it's hard to imagine what garbage it would write there if not, but I didn't read super closely or test in practice.

Reply:
yeah you're right.